crates/adapters/src/format/avro/output.rs

```diff
@@ -0,0 +1,394 @@
+            if let Some(token) = &config.registry_authorization_token {
+                sr_settings_builder.set_token_authorization(token);
+            }
+
+            for (key, val) in config.registry_headers.iter() {
+                sr_settings_builder.add_header(key.as_str(), val.as_str());
+            }
+
+            if let Some(proxy) = &config.registry_proxy {
+                sr_settings_builder.set_proxy(proxy.as_str());
+            }
+
+            if let Some(timeout) = config.registry_timeout_secs {
+                sr_settings_builder.set_timeout(Duration::from_secs(timeout));
+            }
+
+            Some(sr_settings_builder.build().map_err(|e| {
+                ControllerError::invalid_encoder_configuration(
+                    endpoint_name,
+                    &format!("invalid schema registry configuration: {e}"),
+                )
+            })?)
+        } else {
+            None
+        };
+
+        let schema_id = if let Some(sr_settings) = &sr_settings {
+            let supplied_schema = get_supplied_schema(&schema);
+            let name = supplied_schema
+                .name
+                .as_ref()
+                .ok_or_else(|| {
+                    ControllerError::invalid_encoder_configuration(
+                        endpoint_name,
+                        "Avro schema must be of type 'record'",
+                    )
+                })?
+                .clone();
+            let registered_schema =
+                post_schema(sr_settings, name, supplied_schema).map_err(|e| {
+                    ControllerError::encode_error(
+                        endpoint_name,
+                        anyhow!("failed to post Avro schema to the schema registry: {e}"),
+                    )
+                })?;
+            debug!(
+                "avro encoder {endpoint_name}: registered new avro schema with id {}",
+                registered_schema.id
+            );
+            registered_schema.id
+        } else {
+            0
+        };
+
+        let mut buffer = vec![0u8; 5];
+        buffer[1..].clone_from_slice(&schema_id.to_be_bytes());
+
+        Ok(Self {
+            output_consumer,
+            schema,
+            buffer,
+        })
+    }
+}
+
+impl Encoder for AvroEncoder {
+    fn consumer(&mut self) -> &mut dyn OutputConsumer {
+        self.output_consumer.as_mut()
+    }
+
+    fn encode(&mut self, batch: &dyn SerBatchReader) -> AnyResult<()> {
+        let mut cursor = CursorWithPolarity::new(batch.cursor(RecordFormat::Avro)?);
+
+        while cursor.key_valid() {
+            if !cursor.val_valid() {
+                cursor.step_key();
+                continue;
+            }
+            let mut w = cursor.weight();
+
+            if w < 0 {
+                // TODO: we currently only support the "plain" Avro flavor that does not
+                // support deletes.  Other formats, e.g., Debezium will allow deletes.
+                cursor.step_key();
```

Comment:
at least this should be logged, but personally I would prefer an error

Reply:
Sometimes skipping deletes is ok, specifically if the set of keys is stable, and there's a matching insert for each delete, and the destination understands upserts. So I want to still allow that. Logging a warning or an error is fine.  I will add that, but see if I can rate-control those warnings.